### PR TITLE
[UNR-3617] Delay actor hierarchy migration by one frame.

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Actor hierarchy migration happened synchronously through recursive call from USpatialActorChannel::ReplicateActor. While simple, this bypassed the regular replication loop in the net driver.
This caused some pre-replication events to be skipped, namely one which copied actor location to replicated movement data. Symptoms were being pulled back to the actor's origin position after teleport for instance.
This change marks actors for migration instead, in order for them to go through a regular replication frame before being migrated.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -739,9 +743,32 @@ int64 USpatialActorChannel::ReplicateActor()
...
+				Actor->OnAuthorityLost();
+				MigrationVirtualWorkerId = SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
+			}
+		}
+		else if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
 		{
-			const AActor* NetOwner = Actor->GetNetOwner();
+			AActor* NetOwner = [this]
+			{
+				AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:764]: Does it affect the expectation of where GetHierarchyRoot is used elsewhere if the GetHierarchyRoot call returns the Actor itself if it has no owner? Would mean we can reduce these 5 lines to 1

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -739,9 +743,32 @@ int64 USpatialActorChannel::ReplicateActor()
...
+				if(HierarchyRoot)
+				{
+					return HierarchyRoot;
+				}
+				return Actor;
+			}();
+			
 
 			uint64 HierarchyAuthorityReceivedTimestamp = AuthorityReceivedTimestamp;
 			if (NetOwner != nullptr)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:774]: This check doesn't make sense anymore since the lambda returns the Actor itself if the root doesn't exist

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 3]
```diff
@@ -342,4 +344,11 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 	// before the actor holding the position for all the hierarchy, it can immediately attempt to migrate back.
 	// Using this timestamp, we can back off attempting migrations for a while.
 	uint64 AuthorityReceivedTimestamp;
+
+	// Contains the frame at which the actor (or its owner) was seen outside its worker's authoritative region.
+	// Used to delay migration by one frame, to allow Pre-Replication events to take place in the regular replication loop.
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:349]: Can we be more explicit here?

I would prefer this to highlight that:
- Contains the frame at which the Actor hierarchy's was marked to be migrated to another worker.
- Used to delay migration  of all Actors in the hierarchy by one frame. This ensures PreReplication is called for each hierarchy Actor, which would not happen if ReplicateActor is called on the current frame.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 4]
```diff
@@ -342,4 +344,11 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
...
+	uint32 MigrationFrame;
+
+	// WorkerId to migrate to.
+	VirtualWorkerId MigrationVirtualWorkerId = SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:353]: This feels a little bit weird. We make the decision on the target migration worker on the frame before, but ignore any changes to the Actor that could affect migration on the next frame. I think forcing net update on the whole hierarchy is necessary, but we may need to evaluate the migration virtual worker ID on the actual migration frame.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 5]
```diff
@@ -739,9 +743,32 @@ int64 USpatialActorChannel::ReplicateActor()
 	if (NetDriver->LoadBalanceStrategy != nullptr &&
 		NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
 	{
-		if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
+		if (MigrationVirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+		{
+			if(MigrationFrame != NetDriver->ReplicationFrame)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:748]:
```suggestion
			if (MigrationFrame != NetDriver->ReplicationFrame)
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h (hunk 4 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:353]: Is one frame going to make such a difference ? If it would, it means we have some jittering between two zones, and I think we should avoid that by having hysterisis transistions.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 1 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:764]: Mostly used in the LockingPolicy, and the code handles null cases in different ways.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h (hunk 4 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:353]: Actually we should at least take into account if the actor is locked.

Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:355]: I agree that if 1 frame makes a difference, there's other problems, but I feel like this change adds additional potential surprise to migration behaviour. What's the reason for determining the ID in advance? Seems to me like evaluating it when migration actually happens wouldn't be very complicated, but I might be missing something.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 1 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Did some refactor, renamed the current GetHierarchyRoot to GetTopmostOwner.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h (hunk 4 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:355]: We have hysteresis already form the migration backoff time. I still believe we should at least reevaluate the hierarchy authority, even if it's one frame. Currently, it's entirely possible that WhoShouldHaveAuthority returns worker X but on the same frame we set authority intent to worker Y, I don't think this okay to leave open.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 6]
```diff
@@ -765,18 +805,7 @@ int64 USpatialActorChannel::ReplicateActor()
...
-					Actor->Role = ROLE_SimulatedProxy;
-					Actor->RemoteRole = ROLE_Authority;
-
-					Actor->OnAuthorityLost();
-
-					if (NetOwner != nullptr)
-					{
-						ForceReplicateOnActorHierarchy(NetDriver, NetOwner, Actor);
-					}
+					MarkActorHierarchyForMigration(NetDriver, NewAuthVirtualWorkerId, HierarchyRoot);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:808]: If the Actor isn't part of a hierarchy (no owner or child Actors), we can avoid needing to delay at all, which seems preferable

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h [hunk 7]
```diff
@@ -34,6 +34,12 @@ inline AActor* GetHierarchyRoot(const AActor* Actor)
 	return Owner;
 }
 
+inline AActor* GetHierarchyRoot(const AActor* Actor)
+{
+	AActor* TopmostOwner = GetTopmostOwner(Actor);
+	return TopmostOwner ? TopmostOwner : const_cast<AActor*>(Actor);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:40]:
```suggestion
	return TopmostOwner != nullptr ? TopmostOwner : const_cast<AActor*>(Actor);
```

**Comment by ImprobableNic** [comment]
The hysterisis we are missing is about deciding to leave a zone, to avoid WhoShouldHaveAuthority jitter on boundaries.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 6 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:808]: Issue is we don't know for sure, in case of teleport for example, maybe the owner will arrive later and locally resolve to null in the meantime.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 8]
```diff
@@ -456,6 +442,40 @@ void USpatialActorChannel::GetLatestAuthorityChangeFromHierarchy(const AActor* H
 	}
 }
 
+void USpatialActorChannel::MarkActorHierarchyForMigration(USpatialNetDriver* NetDriver, VirtualWorkerId NewAuthVirtualWorkerId, AActor* HierarchyActor)
+{
+	if (HierarchyActor->GetIsReplicated() && HierarchyActor->HasAuthority())
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:447]: What would happen if you try to migrate while waiting for something in your hierarchy to get authority?

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:447]: Or are we just ignoring such cases until someone deals with working sets?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 9]
```diff
@@ -228,6 +212,8 @@ void USpatialActorChannel::Init(UNetConnection* InConnection, int32 ChannelIndex
 	LastPositionSinceUpdate = FVector::ZeroVector;
 	TimeWhenPositionLastUpdated = 0.0f;
 	AuthorityReceivedTimestamp = 0;
+	MigrationFrame = 0;
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:215]: Why does this live in actor channel?
It seems cleaner to have meta-entity stuff not tightly coupled to unreal stuff. Especially as actor channels don't always exist and it's hard to tell what the state of all entities is at this point.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 10]
```diff
@@ -739,14 +759,34 @@ int64 USpatialActorChannel::ReplicateActor()
 	if (NetDriver->LoadBalanceStrategy != nullptr &&
 		NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
 	{
-		if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
+		if (MigrationVirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+		{
+			if (NetDriver->LockingPolicy->IsLocked(Actor))
+			{
+				AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
+				DropActorHierarchyMigration(NetDriver, HierarchyRoot);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:767]: Do we need to migrate them all at once? Does this break if the user changes the locking policy?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h (hunk 4 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:355]: Waiting a tick isn't that different from just not having synchronous load balancing.
Right now I agree you should determine the worker at the last possible moment.
Broadly speaking if we are worried about actors being changed such that the worker they are moving towards can not simulate them at all then we should have workers inspect certain parts of entity state before creating the actor to determine if it can actually simulate it (this can potentially be done using completeness).
The odds are you want to do this anyway if you can have things move into a dungeon phase. Worth noting this also places hard limits on what a user can or should be able to do with locking.

TL;DR this doesn't really introduce new problems but I agree that we should work out what worker is being moved to when it is moved.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 7 above)
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:40]: Is this const casting because the locking policy has the wrong signature? Or do we need to get non const actors from const actors?

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:40]: GetOwner gives you a non const actor pointer.
And from the POV of this function, the given actor is not modified.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 8 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:447]: The case I was seeing was more an actor spawned on another worker trying to "join" its hierarchy. So it should definitely move.
This should be mentioned as a corner case of https://improbableio.atlassian.net/browse/UNR-3539

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:215]: For lack of a better place to put this information. If you have an idea where that could live ?
This is duplicated information amongst all actors of a group, so it could belong to an object representing the group eventually.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 10 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:767]: I expect it to break if another locking policy does not respect ownership.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h (hunk 4 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: Changed to get the destination worker right before migration.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 11]
```diff
@@ -739,47 +764,69 @@ int64 USpatialActorChannel::ReplicateActor()
...
+					const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*HierarchyRoot);
+					if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+					{
+						UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+						DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+					}
+					else if (NewAuthVirtualWorkerId == NetDriver->LoadBalanceStrategy->GetLocalVirtualWorkerId())
+					{
+						DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+					}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:788]: Why do you need these drop hierarchy calls? If the above condition `MigrationVirtualWorkerId.GetValue() == SpatialConstants::INVALID_VIRTUAL_WORKER_ID` is true then we know the whole hierarchy has the same set?

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:788]: So at first, I am marking the hierarchy as needing to be migrated by initializing the TOptional with an invalid worker Id. On the next frame, if the optional is valued and we have an invalid worker Id, it means we have yet to determine the destination, and have encountered the first actor in the hierarchy.

We actually computing the destination, and if it results in an invalid worker Id, or the local worker Id, I drop the attempt to migrate for the whole hierarchy for this frame (no need for the other actors to try the same computation).

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h [hunk 12]
```diff
@@ -44,6 +50,23 @@ inline FString GetConnectionOwningWorkerId(const AActor* Actor)
 	return FString();
 }
 
+template <typename Functor>
+void ForeachReplicatedActorInHierarchy(AActor* HierarchyActor, Functor&& InFunctor)
+{
+	if (HierarchyActor)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:56]:
```suggestion
	if (HierarchyActor != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h [hunk 13]
```diff
@@ -44,6 +50,23 @@ inline FString GetConnectionOwningWorkerId(const AActor* Actor)
...
+	{
+		if (HierarchyActor->GetIsReplicated())
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:58]: If we find a non replicated Actor do we want to continue through its hierarchy?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 14]
```diff
@@ -739,47 +764,69 @@ int64 USpatialActorChannel::ReplicateActor()
 	if (NetDriver->LoadBalanceStrategy != nullptr &&
 		NetDriver->StaticComponentView->HasAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:765]: Not directly a result of your PR, but the whole migrations section has gotten huge, can we extract to its own function?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 11 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:788]: Ah okay, I see, thanks. Do `MigrationFrame != NetDriver->ReplicationFrame` and `MigrationVirtualWorkerId.IsSet()` represent the same state then? I.e. we change the top level conditional for the section to `MigrationFrame != NetDriver->ReplicationFrame` then MigrationVirtualWorkerId doesn't need to be optional?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 15]
```diff
@@ -342,4 +345,14 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 	// before the actor holding the position for all the hierarchy, it can immediately attempt to migrate back.
 	// Using this timestamp, we can back off attempting migrations for a while.
 	uint64 AuthorityReceivedTimestamp;
+
+	// Contains the frame at which the actor hierarchy was marked to be migrated to another worker.
+	// Used to delay the hierarchy's migration by one frame, to allow all actors in the hierarchy to go through
+	// the regular replication loop (to run Pre-Replication events for instance) before being migrated together.
+	uint32 MigrationFrame;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:352]: I think there might be a tiny bit of ambiguity for someone debugging since this is actually the frame when migration was initiated, rather than the actual migration frame as it literally implies. I know the comment clarifies, but can we also change the name to something like `FrameWhenActorHierarchyMigrationStarted`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 6 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: I don't think this is true? I think there's a difference between the owner and child references being unresolved or null?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 13 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:58]: Maybe ? I could imagine a convoluted setup where an actor has some non-replicated actor limbs, which are locally and deterministically animated, but is able to pick up replicated things from the ground.
While I'm not sure it would remain a valid setup for long, I do not want to actually break this scenario.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 16]
```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+	{
+		AActor* HierarchyRoot = SpatialGDK::GetHierarchyRoot(Actor);
+
+		// If we were locked in between replication frames, drop migration attempt.
+		if (NetDriver->LockingPolicy->IsLocked(Actor))
+		{
+			DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+		}
+		// Are we in a different frame from the one we determined we should try to migrate ?
+		else if (MigrationInitiatedFrame != NetDriver->ReplicationFrame)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:497]: ReplicationFrame isn't used when running with the replication graph, the rep graph has it's own counter.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h [hunk 17]
```diff
@@ -44,6 +50,23 @@ inline FString GetConnectionOwningWorkerId(const AActor* Actor)
...
+		{
+			InFunctor(HierarchyActor);
+		}
+
+		for (AActor* Child : HierarchyActor->Children)
+		{
+			ForeachReplicatedActorInHierarchy(Child, InFunctor);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:65]: Presumably this should forward the functor

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 6 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Yes, but that is not an information we can retrieve by merely checking the owner pointer. We would have to visit the properties involved in the hierarchy, and check if we recorded some UnrealObjectRef for them.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 11 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: They are not the same state.
MigrationWorkerVirtualId.IsSet() means "someone in the hierarchy has requested a migration on this frame, or on a previous frame".
MigrationFrame != NetDriver->ReplicationFrame means that we are at a different replication frame than the one where we determined we should migrate.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h (hunk 13 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialActorUtils.h:58]: Cool, let's just make the case for everything then 👍

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 11 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: My point is that you can replace `MigrationWorkerVirtualId.IsSet()` with  `MigrationFrame != NetDriver->ReplicationFrame` on the top level conditional and make `MigrationWorkerVirtualId` not be an optional anymore?

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: I think this conversation is hidden cos outdated so moving to the correct place

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 18]
```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+		{
+			Channel->MigrationVirtualWorkerId.Reset();
+		}
+	}));
+}
+
+void USpatialActorChannel::HandleActorLoadBalancing()
+{
+	// if the field is set it means we should try to migrate authority to another worker
+	if (MigrationVirtualWorkerId.IsSet())
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:487]: I think you can replace `MigrationWorkerVirtualId.IsSet()` with `MigrationFrame != NetDriver->ReplicationFrame` and make MigrationWorkerVirtualId not be an optional anymore?

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:487]: Replied in the previous comment-->

They are not the same state.
MigrationWorkerVirtualId.IsSet() means that someone in the hierarchy has requested a migration on this frame, or on a previous frame.
MigrationFrame != NetDriver->ReplicationFrame means that we are at a different replication frame than the one where we determined we should migrate, and at this frame, all actors of the hierarchy will go through replication.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 19]
```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+
+		// If we were locked in between replication frames, drop migration attempt.
+		if (NetDriver->LockingPolicy->IsLocked(Actor))
+		{
+			DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+		}
+		// Are we in a different frame from the one we determined we should try to migrate ?
+		else if (MigrationInitiatedFrame != NetDriver->ReplicationFrame)
+		{
+			// An invalid worker id means we encounter the first actor in the hierarchy, which will determine where all the other actors will migrate.
```
Comment (KeyMaster-) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:499]: I don't think this comment is correct anymore - the migration target is now determined by the hierarchy root

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 20]
```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+			if (MigrationVirtualWorkerId.GetValue() == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
 			{
-				OutTimestamp = Channel->AuthorityReceivedTimestamp;
+				const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*HierarchyRoot);
+				if (NewAuthVirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+				{
+					UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
```
Comment (KeyMaster-) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:505]: We got the invalid worker ID from the hierarchy root, so the log should use its name.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 21]
```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+				{
+					DropActorHierarchyMigration(NetDriver, HierarchyRoot);
+				}
+				else
+				{
+					SetActorHierarchyMigrationDestination(NetDriver, NewAuthVirtualWorkerId, HierarchyRoot);
+				}
+			}
+
+			if (const VirtualWorkerId NewAuthVirtualWorkerId = MigrationVirtualWorkerId.GetValue())
```
Comment (KeyMaster-) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:518]: What's the reason for declaring `NewAuthVirtualWorkerId`? It's not used in the block. Also, I'm assuming this is supposed to check for whether `MigrationVirtualWorkerId` is valid, in which case I'd expect it to be compared against `INVALID_VIRTUAL_WORKER_ID`.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 22]
```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+				Actor->OnAuthorityLost();
+				MigrationVirtualWorkerId.Reset();
 			}
 		}
 	}
+	// Checking if we should retain authority.
+	else if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
+	{
+		AActor* TopmostOwner = SpatialGDK::GetTopmostOwner(Actor);
+		AActor* HierarchyRoot = TopmostOwner ? TopmostOwner : Actor;
```
Comment (KeyMaster-) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:535]: How come that the method to get `HierarchyRoot` is different here vs line 489? (`SpatialGDK::GetHierarchyRoot` vs this)

**Comment by ImprobableNic** [comment]
Closed in favor of a cleaner approach in https://github.com/spatialos/UnrealGDK/pull/2302
